Gutenberg Plugin: Add hook to allow `writing-mode` as a safe CSS property

t-hamano opened the pull request:
Fixes #54344
Alternative to #54345
For more information, see [this comment](https://github.com/WordPress/gutenberg/pull/54345#pullrequestreview-1630084109)

## What?

This PR includes the following three:

- Added `safe_style_css` filter to prevent `wiriting-mode` property from being removed by `safecss_filter_attr()` function
- Remove explicit addition of `writing-mode` style in render callback function of the Post Navigation Link block which is no longer needed
- Updated the Style Engine testing

## Why?

For dynamic blocks, as with other block support styles, there is no need to explicitly pass the style, as it should be automatically output by the `get_wrapper_attributes()` function.

However, this property is not yet allowed in the core, so we need to allow it in the Gutenberg plugin first.

## How?

`writing-mode` (text orientation) is supported in WordPress 6.3. Therefore, this hook originally needed to be added in the `lib/compat/wordpress-6.3` directory. However, this property will only be allowed in core WordPress 6.4 and above, so I added the hook to the `lib/compat/wordpress-6.4` directory.

## Testing Instructions

To test this PR, enable Emptytheme and update theme.json as below:

```json
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"settings": {
		"appearanceTools": true,
		"layout": {
			"contentSize": "840px",
			"wideSize": "1100px"
		},
		"typography": {
			"writingMode": true
		}
	}
}
```

### Footnotes Block

- Add some footnotes to the paragraph block.
- In the Typography panel of the Footnotes block, change orientation to "vertical".
- In the front end, the block should be displayed vertically.

### Post Navigation Block

- Create three posts.
- Insert a Previous Post block and a Next Post block in the middle post.
- In the Typography panel, change orientation to "vertical".
- In the front end, the block should be displayed vertically.

t-hamano committed:
Gutenberg Plugin: Add hook to allow `writing-mode` as a safe CSS property

carolinan commented:
As far as I know, it was never supported in 6.3. I don't know if this was a decision made or if the backport just did not get done.

carolinan commented:
Another question is if we want to add support for text-orientation at the same time or if this needs to wait until later:
https://github.com/WordPress/gutenberg/pull/52624

t-hamano commented:
> As far as I know, it was never supported in 6.3. I don't know if this was a decision made or if the backport just did not get done.

Sorry, this was my mistake. Indeed, writing mode seems to be supported starting with WordPress 6.4.

https://core.trac.wordpress.org/ticket/59306

> Another question is if we want to add support for text-orientation at the same time or if this needs to wait until later:

Personally, I think it would be better to merge this PR first, which has fewer files to add/update. That way, #52624 would only need to update the `safe_style_css` hook and style engine test after rebasing with the latest trunk.

andrewserong commented on packages/block-library/src/post-navigation-link/index.php:
```diff
@@ -28,14 +28,9 @@ function render_block_core_post_navigation_link( $attributes, $content ) {
-		$styles = "writing-mode: {$attributes['style']['typography']['writingMode']};";
-	}
 	$wrapper_attributes = get_block_wrapper_attributes(
 		array(
 			'class' => $classes,
-			'style' => $styles,
 		)
 	);
```
Since this change will be bundled in with the packages update for core, would it be worth trying to land the change to core's `safe_style_css` list before attempting to land this Gutenberg PR?

t-hamano replied:
I see, I think that's certainly true. 

Would it be ideal to wait to merge this PR until a patch has been submitted to the core that adds `writing-mode` to the `safe_style_css` list and is committed in the core?

andrewserong replied:
I don't think it necessarily needs to be committed to core before merging this PR, but I'd recommend opening a trac ticket and corresponding PR for `wordpress-develop` before merging if you have the time, that way the change will be ready to land in time.

andrewserong reviewed (approved):
This is working nicely for me, thanks for adding it in!

✅ Writing-mode now outputs correctly for the footnotes block on the site frontend
✅ With this change the writing mode still outputs correctly for the post navigation link block

<img width="481" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/c254dad0-a09f-4f22-9771-0081ad31e2a0">

Just left a comment about backporting to core. If this PR lands for 6.4, it'd be good to ensure the backport for core lands as well so that the adjustment to the post navigation link block doesn't cause any regressions.

LGTM! ✨

t-hamano commented:
Added a ticket and a patch to the core: https://core.trac.wordpress.org/ticket/59387

Also added to the backport tracking issue #54177.

I think everything is ready now, so let's merge this PR 👍

t-hamano commented:
@carolinan Could you rebase #52624 with the latest trunk and add `text-orientation` to the hook in `lib/compat/wordpress-6.4/kses.php`?

carolinan commented:
> @carolinan Could you rebase #52624 with the latest trunk and add `text-orientation` to the hook in `lib/compat/wordpress-6.4/kses.php`?

Done
